File: tests/sample.php
<?php

declare(strict_types=1);

readonly class User
{
    public function __construct(
        public int    $id,
        public string $name,
        public string $email,
    ) {}
}

class UserService
{
    private array $cache = [];

    public function getUser(int $id): ?User
    {
        if (isset($this->cache[$id])) {
            return $this->cache[$id];
        }

        $data = $this->fetchFromApi($id);
        if ($data === null) return null;

        $user = new User($data['id'], $data['name'], $data['email']);
        $this->cache[$id] = $user;
        return $user;
    }

    private function fetchFromApi(int $id): ?array
    {
        $json = file_get_contents("https://api.example.com/users/{$id}");
        if ($json === false) return null;
        return json_decode($json, true);
    }
}

$service = new UserService();
$user = $service->getUser(1);
echo $user ? "Hello, {$user->name}!" : "User not found";

File: tests/sample.rb
require "net/http"
require "json"

User = Data.define(:id, :name, :email)

class UserService
  def initialize
    @cache = {}
  end

  def get_user(id)
    @cache[id] ||= fetch_from_api(id)
  end

  private

  def fetch_from_api(id)
    uri = URI("https://api.example.com/users/#{id}")
    response = Net::HTTP.get_response(uri)
    return nil unless response.is_a?(Net::HTTPSuccess)

    data = JSON.parse(response.body, symbolize_names: true)
    User.new(**data)
  rescue StandardError => e
    warn "Failed to fetch user #{id}: #{e.message}"
    nil
  end
end

svc = UserService.new
if (user = svc.get_user(1))
  puts "Hello, #{user.name}!"
else
  puts "User not found"
end
